framework/src/components/system/system.js

```diff
@@ -0,0 +1,164 @@
+ * - nethash
+ * - broadhash
+ * - nonce
+ *
+ * @class
+ * @requires crypto
+ * @requires os
+ * @requires semver
+ * @param {Object} options - System options
+ * @param {Object} logger
+ */
+class System {
+	constructor(config, logger, storage) {
+		this.logger = logger;
+		this.storage = storage;
+		this.headers = {
+			os: os.platform() + os.release(),
+			version: config.version,
+			minVersion: config.minVersion,
+			protocolVersion: config.protocolVersion,
+			height: 1,
+			nethash: config.nethash,
+			broadhash: config.nethash,
+			nonce: config.nonce,
+		};
+	}
+
+	/**
+	 * Checks nethash (network) compatibility.
+	 *
+	 * @param {string} nethash
+	 * @returns {boolean}
+	 */
+	networkCompatible(nethash) {
+		if (!nethash) {
+			return false;
+		}
+		return this.headers.nethash === nethash;
+	}
+
+	/**
+	 * Checks version compatibility from input param against private values.
+	 *
+	 * @param {string} version
+	 * @returns {boolean}
+	 */
+	versionCompatible(version) {
+		if (!version) {
+			return false;
+		}
+		return semver.gte(version, this.headers.minVersion);
+	}
+
+	/**
+	 * Checks protocol version compatibility from input param against
+	 * private values.
+	 *
+	 * @param protocolVersion
+	 * @returns {boolean}
+	 */
+	protocolVersionCompatible(protocolVersion) {
+		if (!protocolVersion) {
+			return false;
+		}
+		const peerHard = parseInt(protocolVersion[0]);
+		const myHard = parseInt(this.headers.protocolVersion[0]);
+		return myHard === peerHard && peerHard >= 1;
+	}
+
+	/**
+	 * Checks nonce (unique app id) compatibility- compatible when different than given.
+	 *
+	 * @param nonce
+	 * @returns {boolean}
+	 */
+	nonceCompatible(nonce) {
+		if (!nonce) {
+			return false;
+		}
+		return nonce && this.headers.nonce !== nonce;
+	}
+
+	/**
+	 * It calculates and returns broadhash.
+	 *
+	 * @param {Error} err
+	 * @returns {string} broadhash
+	 */
+	getBroadhash(cb) {
+		if (typeof cb !== 'function') {
+			return this.headers.broadhash;
+		}
+		return this.storage.entities.Block.get(
+			{},
+			{
+				limit: 5,
+				sort: 'height:desc',
+			}
+		)
+			.then(blocks => {
+				if (blocks.length <= 1) {
+					// In case that we have only genesis block in database (query returns 1 row) - skip broadhash update
+					return setImmediate(cb, null, this.headers.nethash);
+				}
+				const seed = blocks.map(row => row.id).join('');
+				const broadhash = crypto
+					.createHash('sha256')
+					.update(seed, 'utf8')
+					.digest()
+					.toString('hex');
+				return setImmediate(cb, null, broadhash);
+			})
+			.catch(err => {
+				this.logger.error(err.stack);
+				return setImmediate(cb, err);
+			});
+	}
+
+	/**
+	 * Updates private broadhash and height values.
+	 *
+	 */
+	async update() {
+		const hash = this.getBroadhash();
+		this.headers.broadhash = hash;
+		const blocks = await this.storage.entities.Block.get(
+			{},
+			{
+				limit: 5,
+				sort: 'height:desc',
+			}
+		);
+		this.headers.height = blocks[0].height;
+	}
```

Comment:
I am not sure if the `headers` are something shared among different modules. When we have seperate `network` module, it will contain its now implementation of headers or related items. 

For now we can proceed with it, but later we should refactor this part to remove this dependency of network module. 

In summary any logic which is useable to share among different modules will come in the system components. e.g. In future we can put `slots` helpers in here. Because different modules may need this information. 

Related to broadhash, we should update broad hash whenever we system receive in block event. Not necessary to generate broadhash on every call. But that can also and should refactored separately.